File: cpp/epi/search_for_student.cc
#include <algorithm>
#include <vector>
#include "fmt_print.h"
#include "generic_test.h"
#include "serialization_traits.h"
using std::string;
using std::vector;

struct Student {
  string name;
  float gpa;
};

template <>
struct SerializationTraits<Student>
    : UserSerTraits<Student, string, float> {
  
  static std::vector<std::string> GetMetricNames(const std::string& arg_name) {
    return {};
  }

  static std::vector<int> GetMetrics(const Student& student) {
    return {};
  }
};

bool operator==(const Student& lhs, const Student& rhs) {
  return (lhs.name == rhs.name) && (lhs.gpa == rhs.gpa);
}

std::ostream& operator<<(std::ostream& out, const Student& student) {
  return PrintTo(out, std::make_tuple(student.name, student.gpa));
}

int FindStudent(const vector<Student>& students, const Student& key) {
  return 0;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::string func_name = "FindStudent";
  std::vector<std::string> param_names{"students", "key"};

  return GenericTestMain(args, "search_for_student.cc", "../test_data/epi/search_for_student.tsv",
                         &FindStudent, DefaultComparator{}, func_name, param_names);
}

File: java/epi/SearchForStudent.java
package epi;
import test_framework.EpiTest;
import test_framework.EpiUserType;
import test_framework.GenericTest;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class SearchForStudent {

  @EpiUserType(ctorParams = {String.class, Float.class})
  public static class Student {
    public String name;
    public Float gpa;

    public Student(String name, Float gpa) {
      this.name = name;
      this.gpa = gpa;
    }

    @Override
    public String toString() {
      return "[" + "name: " + name + ", gpa: " + gpa + "]";
    }
  }

  @EpiTest(testDataFile = "../test_data/epi/search_for_student.tsv")
  public static int findStudent(List<Student> A, Student key) {
    int res = Collections.binarySearch(A, key, new Comparator<Student>() {
      @Override
      public int compare(Student a, Student b) {
        if (b.gpa.compareTo(a.gpa) != 0) {
          return b.gpa.compareTo(a.gpa);
        }

        return a.name.compareTo(b.name);
      }
    });

    if (res < 0) {
      return -1;
    }
    return res;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "SearchForStudent.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
